src/hmem_synapseai.c

```diff
@@ -43,43 +43,99 @@
 #include "habanalabs/synapse_api.h"
 
 struct synapseai_ops {
+	synStatus (*synInitialize)();
+	synStatus (*synDestroy)();
+	synStatus (*synDeviceGetCount)(uint32_t* pCount);
 	int (*hcclLookupDMABuff)(uint64_t addr, uint64_t size, int* fd);
 };
 
 static void *synapseai_handle;
 static struct synapseai_ops synapseai_ops;
 
-/**
- * @brief Initialize SynapseAI hmem interface by dlopen the SynapseAI library
- * 
- * @return On success, return 0. On failure, return negative error code.
- */
-int synapseai_init(void)
-{
+int synapseai_dl_init() {
+	int err = FI_SUCCESS;
 	synapseai_handle = dlopen("libSynapse.so", RTLD_NOW);
 	if (!synapseai_handle) {
 		FI_INFO(&core_prov, FI_LOG_CORE,
 			"Failed to dlopen libSynapse.so\n");
-		return -FI_ENOSYS;
+		err = -FI_ENOSYS;
+		goto out;
+	}
+
+	synapseai_ops.synInitialize = dlsym(synapseai_handle, "synInitialize");
+	if (!synapseai_ops.synInitialize) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to find synInitialize\n");
+		err = -FI_ENOSYS;
+		goto out;
+	}
+
+	synapseai_ops.synDestroy = dlsym(synapseai_handle, "synDestroy");
+	if (!synapseai_ops.synDestroy) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to find synDestroy\n");
+		err = -FI_ENOSYS;
+		goto out;
+	}
+
+	synapseai_ops.synDeviceGetCount = dlsym(synapseai_handle, "synDeviceGetCount");
+	if (!synapseai_ops.synDeviceGetCount) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to find synDeviceGetCount\n");
+		err = -FI_ENOSYS;
+		goto out;
 	}
 
 	synapseai_ops.hcclLookupDMABuff = dlsym(synapseai_handle, "hcclLookupDMABuff");
 	if (!synapseai_ops.hcclLookupDMABuff) {
 		FI_WARN(&core_prov, FI_LOG_CORE,
 			"Failed to find hcclLookupDMABuff\n");
-		goto err;
+		err = -FI_ENOSYS;
+		goto out;
 	}
 
```

Comment:
Shouldn't you return FI_SUCCESS before reaching `out`? Otherwise the dl handle will be closed